### ISSUE TITLE
Suggestion: Providing a C++ API as part of sources/android/ndk_helper

### ISSUE THREAD

**Issue by pjmlp** [android-ndk/ndk#476]
#### Description

Picking up on the AMA discussion about [C++ APIs](https://www.reddit.com/r/androiddev/comments/6nuq0a/were_on_the_engineering_team_for_android_o_ask_us/dkfxx32/), just wanted to leave my thoughts about it here.

Maybe extend the set of classes available under _sources/android/ndk_helper_ with wrappers for all
platform APIs described in https://developer.android.com/ndk/reference/index.html.

The code would be at very least C++14, or even better C++17, given the actual inclusion of clang 5.x in the current NDK releases.

Taking advantage of modern C++ best practices, maybe with conditional compilation for RTTI and exceptions support

#### Examples

A couple of basic examples:

### ASharedMemory

```C++
namespace android::memory  {
    class SharedMemory final {
    public:
        SharedMemory (const std::string& name, size_t size);
        SharedMemory (std::string_view name, size_t size);
        ~SharedMemory();  // Use RAII to release the memory
 
        size_t getSize ();
        int setProt (int prot)
    };
}
```

### ATrace

```C++
namespace android::trace  {
    class Trace  final {
    public:
        Trace  (const std::string& sectionName);
        Trace  (std::string_view sectionName);
        ~Trace  (); // Use RAII to close trace section
 
        static bool isEnabled() noexcept;
        static void beginSection(const std::string& sectionName) noexcept;
        static void beginSection(std::string_view sectionName) noexcept;
        static void endSection() noexcept;
    };
}
```

This way one can invoke the static member functions C style, but with string safety, or just
declare a code block with a _Trace_ stack instance for the code segment being debugged.

### Bitmap

```C++
namespace android::graphics   {
    class Bitmap final {
    public:
        Bitmap (JNIEnv *env, jobject jbitmap);
        ~Bitmap ();  // Use RAII to release the bitmap/unlock it
 
        int getInfo (AndroidBitmapInfo& info);
        int lockPixels (void **addrPtr);
        int unlockPixels ();

        // other nice to have, 2D drawing member functions
    };
}
```

#### Environment Details

Not all of these will be relevant to every bug, but please provide as much
information as you can.

* NDK Version: 15 or higher
* Build sytem: all NDK supported systems
* Host OS: Windows, any
* Compiler: clang 5.0 or higher.
* ABI: all
* STL: yes

**Comment by DanAlbert** [comment]
We're going to play with adding these directly to the sysroot rather than adding them to ndk_helper, actually. We'll see how that goes. My guess is that some of them will be more complicated than something that can reasonably be defined entirely inline in the headers and will need an actual library, but the first API that's actually tried this approach hasn't needed it.

For most of the APIs we're most likely going to end up refiling these bugs against their owner, but we'll use this as a tracking bug.

Anyone have any APIs they *really* want this for? There are a lot and it would be nice to have some idea of priority.

**Comment by pjmlp** [comment]
On my specific case, hobby coding for Android doing graphics related stuff, C++ support for Canvas and BitmapFactory would be nice.

Like everyone else, I just get to use SDL and  external image loading libraries as workaround.

Given the already existing support on Android for such features, it goes a bit against the advocacy of having smaller size APKs.

**Comment by Zingam** [comment]
Is this about wrapping Android's Java platform APIs in C++ interfaces? 
If so my guess is that the NDK is used primarily for games development then all APIs that are relevant to game development should be prioritized. One should be able to develop a complete game without touching Java.

**Comment by DanAlbert** [comment]
> Is this about wrapping Android's Java platform APIs in C++ interfaces?

Adding C++ wrappers to the C NDK APIs, actually.

**Comment by DanAlbert** [comment]
We'll have the first prototype of this approach available Soon™ (I'd be more specific if I could, but the frameworks folks are still working on it aiui; probably will be in r17 though).

Once that is done I think we'll split this bug up into one per C API header rather than try to track them all on one, but this will remain open until we at least get a POC shipped.

**Comment by pjmlp** [comment]
Nice this is being picked up.

In case you also need feedback from hobby NDK users, I will gladly provide it.

**Comment by DanAlbert** [comment]
To make this a bit less nebulous, I've forked this out into three separate bugs for the APIs suggested here:

https://github.com/android-ndk/ndk/issues/820

**Comment by twaik** [comment]
It would be cool to access some Activity's feature like this: 
https://github.com/twaik/termux-api/blob/master/termux-api-so/libjni/android/app/Activity.cpp
https://github.com/twaik/termux-api/blob/master/termux-api-so/libjni/android/content/Context.cpp
https://github.com/twaik/termux-api/blob/master/termux-api-so/libjni/android/content/Intent.cpp
i.e. when you trying to design some app designed not to use Java. 

I was trying to implement it but I didn't finish it.

**Comment by DanAlbert** [comment]
Please file new bugs for the APIs you want. We're not app developers, so please include a reference to either the Java or NDK API you're referring to.

Locking this thread to encourage people to file individual API bugs.
